### ISSUE TITLE
ClientId-based authentication causes a crash in App Service with MSI enabled

### ISSUE THREAD

**Issue by rynowak** [Azure/go-autorest#547]
## Summary

I happened on a case where my application works locally in a container, but does not work in app service containers. It turns out the delta is whether MSI is enabled on app service, which causes the adal library to behave differently when used with non-MSI authentication. I am using Client-ID/Client-Secret auth (not MSI).

I was able to find a workaround. Disabling MSI for my app service site fixes the issue. 

Again, I want to make it clear that I am not using MSI for auth, but having a managed identity assigned in App Service breaks stuff.

## Repro Code

You should be able to reproduce this in a console application by setting the `MSI_ENDPOINT` or `MSI_SECRET` environment variables

```go
clientcfg := auth.NewClientCredentialsConfig(os.Getenv("CLIENT_ID"), os.Getenv("CLIENT_SECRET"), os.Getenv("TENANT_ID"))
auth, err := clientcfg.Authorizer()
if err != nil {
    return nil, err
}

token, err := clientcfg.ServicePrincipalToken()
if err != nil {
    return nil, err
}

err = token.EnsureFresh()
if err != nil {
    return nil, err
}
```

The failure looks like this:

```
2020-08-11T01:40:22.052723200Z panic: runtime error: invalid memory address or nil pointer dereference
2020-08-11T01:40:22.054713623Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x126651d]
2020-08-11T01:40:22.055131728Z
2020-08-11T01:40:22.055508233Z goroutine 1 [running]:
2020-08-11T01:40:22.056591545Z github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).refreshInternal(0xc00024b200, 0x1817e80, 0xc00004a038, 0x1588a0a, 0x1d, 0x0, 0x0)
2020-08-11T01:40:22.059709781Z 	/go/pkg/mod/github.com/!azure/go-autorest/autorest/adal@v0.9.1/token.go:928 +0x6dd
2020-08-11T01:40:22.061509902Z github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).EnsureFreshWithContext(0xc00024b200, 0x1817e80, 0xc00004a038, 0x0, 0x0)
2020-08-11T01:40:22.062163509Z 	/go/pkg/mod/github.com/!azure/go-autorest/autorest/adal@v0.9.1/token.go:794 +0x18d
2020-08-11T01:40:22.062722416Z github.com/Azure/go-autorest/autorest/adal.(*ServicePrincipalToken).EnsureFresh(...)
2020-08-11T01:40:22.063512125Z 	/go/pkg/mod/github.com/!azure/go-autorest/autorest/adal@v0.9.1/token.go:782
2020-08-11T01:40:22.068735885Z github.com/Azure/azure-oam-curp/pkg/curp/k8sauth.CreateClient(0x0, 0x0, 0x0)
2020-08-11T01:40:22.069769797Z 	/src/pkg/curp/k8sauth/client.go:56 +0x4f8
```

I was able to investigate this and find a few issues that are contributing to the bad error message. 

- [isIMDS](https://github.com/Azure/go-autorest/blob/master/autorest/adal/token.go#L881) returns true for this case because of the logic in `isAppService` despite the fact that it should be a normal OAuth call
- [retryForIMDS](https://github.com/Azure/go-autorest/blob/master/autorest/adal/token.go#L960 - can sometimes return `nil, nil` (no error but no response). I did not debug the exact conditions that precipitate that.
- [refreshInternal](https://github.com/Azure/go-autorest/blob/master/autorest/adal/token.go#L928) then gets a `nil` http response pointer, and segfaults

**Comment by jhendrixMSFT** [comment]
Thanks for the details, I believe this is also the reason for the crash described in https://github.com/terraform-providers/terraform-provider-azurerm/issues/6773.  I have suspected that `retryForIMDS()` was returning `nil` for the response and error but had no confirmation from the other report.

**Comment by jhendrixMSFT** [comment]
`resp` is nil because `MaxMSIRefreshAttempts` on the `ServicePrincipalToken` type is zero, which is expected since you aren't really using MSI authentication.

**Comment by jhendrixMSFT** [comment]
This is fixed in autorest/adal/v0.9.2
